posix: semaphore: implement named semaphore functions

ycsin opened the pull request:
Implemented:
- `sem_open()`
- `sem_close()`
- `sem_unlink()`

Fixes #66955
Fixes #66956
Fixes #66957

ycsin committed:
posix: semaphore: implement `sem_open()`, `sem_unlink()` & `sem_close()`

Implements `sem_open()`, `sem_unlink()` & `sem_close()`
functions and added tests for them.

Updated existing tests and POSIX docs.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
MAINTAINERS: add myself as a POSIX collaborator

Add myself as a collaborator in the POSIX subsystem.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

jukkar commented on lib/posix/semaphore.c:
```diff
@@ -1,11 +1,76 @@
+};
+
+/* Initialize the list */
+sys_slist_t sem_list = SYS_SLIST_STATIC_INIT(&sem_list);
```
Shouldn't this be `static`

ycsin replied:
fixed this

jukkar commented on lib/posix/semaphore.c:
```diff
@@ -1,11 +1,76 @@
+/* Initialize the list */
+sys_slist_t sem_list = SYS_SLIST_STATIC_INIT(&sem_list);
+
+K_MUTEX_DEFINE(nsem_mutex);
```
and this one too

ycsin replied:
and this, thanks!

jukkar commented on lib/posix/semaphore.c:
```diff
@@ -1,11 +1,76 @@
+	sys_snode_t snode;
+	sem_t sem;
+	atomic_t ref_count;
+	char *name;
```
We could avoid malloc if allocated like this
```
char name[CONFIG_SEM_NAMELEN_MAX];
```
I think we have this pattern in many places in the code space.

ycsin replied:
I'm gonna gather more comments for this one, doing so will allocate a fixed sized array regardless of the actual length of the name.

Alternatively we can allocate the buffer for the `nsem_obj` and the `name` array in one go and the total memory allocated will depend on the name length. Ideally we would be able to `k_realloc` the buffer once the sem is unlinked, however the `k_realloc` isn't available (#41151), so I opted for the current implementation as the name buffer can be freed separately upon `sem_unlink`.

cfriedt replied:
I would say a good follow-up PR would be to add dynamic allocation support.

The existing pattern that doesn't use malloc should be fine for filling the immediate gap. Thanks for your help ❤️

jukkar commented on lib/posix/semaphore.c:
```diff
@@ -1,11 +1,76 @@
+
+K_MUTEX_DEFINE(nsem_mutex);
+
+#if defined(__sparc__)
```
Instead of directly depend on board, perhaps depend on a config that is only set for needed board.

ycsin replied:
@cfriedt I simply copied this from `mqueue.c`, should we place this somewhere more generic?

jukkar replied:
My comment is related to the fact that we imho should have checks depending on features and not directly boards. The board config can then enable them, or we could enable the features by default for certain boards.

cfriedt replied:
Definitely the malloc aspect is nice to have @ycsin.

I actually would really love to apply a simply pattern to all of the "pooled" types in posix, and we actually really also need this to separate pthreads / thread stacks / c11 threads, etc.

In any case though, solve that pattern once and then it can be used over and over again in posix, for sure.

What about this convention for mixing static and dynamic storage for "pooled types", @jukkar ?

* If the size is negative (Kconfig allows negative ints)
  - the absolute value of the size (minus one) is used for number of static allocations in the pool
  - the negative bit indicates that items can also be allocated from the heap
* If it's positive only, then that would mean only static allocations

ycsin replied:
Fixed and rebased

jukkar commented on lib/posix/Kconfig.semaphore:
```diff
@@ -8,3 +8,10 @@ config SEM_VALUE_MAX
+
+config SEM_NAMELEN_MAX
+	int "Maximum name length"
+	default 16
```
BTW, regarding the pointer vs array tradeoffs. If we set the default to 6, then there would be no penalty when comparing dynamic and static allocation sizes. Anyway, this was just a note, no need to change this.

cfriedt replied:
It might be good to note in the `help` section, is that the value 255 corresponds to `NAME_MAX`.

ycsin replied:
> It might be good to note in the `help` section that the value 255 corresponds to `NAME_MAX`.

Fixed

cfriedt commented on lib/posix/semaphore.c:
```diff
@@ -148,3 +213,159 @@ int sem_wait(sem_t *semaphore)
+sem_t *sem_open(const char *name, int oflags, ...)
+{
+	va_list va;
+	mode_t mode;
```
I remember this was an issue on `__sparc__` so yea, definitely valid.

Instead of `mode_t` here, you might need to use `PROMOTED_MODE_T`, but it's hard to say for me. Look to see who made the commit that added it, and add them to review.

In any case, this is a really cool feature. It kind of gets you thinking about other things, like mmap, and pipe to name a couple.

cfriedt commented on doc/services/portability/posix/option_groups/index.rst:
```diff
@@ -367,14 +367,14 @@ POSIX_SEMAPHORES
    :widths: 50,10
 
-    sem_close(),
+    sem_close(),yes
```
Just one minor nit - the doc changes should probably be in a separate commit.

ycsin replied:
fixed

cfriedt commented on MAINTAINERS.yml:
```diff
@@ -2728,6 +2728,8 @@ POSIX API layer:
   maintainers:
     - cfriedt
+  collaborators:
+    - ycsin
```
🥳

cfriedt commented on lib/posix/semaphore.c:
```diff
@@ -148,3 +213,159 @@ int sem_wait(sem_t *semaphore)
+		}
+
+		va_start(va, oflags);
+		mode = va_arg(va, PROMOTED_MODE_T);
```
```suggestion
		BUILD_ASSERT(sizeof(mode_t) <= sizeof(int));
		mode = va_arg(va, int);
```

ycsin replied:
fixed & rebased

cfriedt commented on tests/posix/common/src/semaphore.c:
```diff
@@ -90,41 +84,189 @@ ZTEST(posix_apis, test_semaphore)
+	zassert_equal(get_sem_list_len(), num_sem);
+
+	/* Name exceeds CONFIG_SEM_NAMELEN_MAX */
+	sem1 = sem_open("0123456789abcdefg", 0, 0, 0);
```
It would be best to modify this so that it uses a (local scope) `static char [CONFIG_SEM_NAMELEN_MAX + 2]`, then fill the string with `a` or something (just remember to terminate with `\0`).

Then it's robust against changes to `CONFIG_SEM_NAMELEN_MAX`.

ycsin replied:
fixed

ycsin committed:
doc: posix: mark `sem_open`, `sem_close` & `sem_unlink` as supported

The `sem_open()`, `sem_close()` & `sem_unlink()` functions are
now implemented, so mark them as supported.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

npitre commented:
This code is racy.

```
sem_t *sem_open(const char *name, int oflags, ...)
{
       /* Check if the named semaphore exists */
       k_mutex_lock(&nsem_mutex, K_FOREVER);
       nsem = find_nsem(name);
       k_mutex_unlock(&nsem_mutex);

       /* !!! preemption point !!! */
       [...]

       if (nsem == NULL) {
                [...]
       } else {
               atomic_inc(&nsem->ref_count);
       }
```

When `find_nsem()` returns a valid non-null pointer for `nsem`,
nothing prevents another thread calling `sem_unlink()` and `sem_close()`
at the indicated preemption point, turning `nsem` into a zombie pointer 
and corrupting freed and potentially reused memory.

cfriedt commented on lib/posix/semaphore.c:
```diff
@@ -148,3 +202,160 @@ int sem_wait(sem_t *semaphore)
+		value = va_arg(va, unsigned int);
+		va_end(va);
+
+		nsem = k_calloc(1, sizeof(struct nsem_obj));
+		if (nsem == NULL) {
+			errno = ENOSPC;
+			goto error;
+		}
+
+		/* goto `cleanup` past this point to avoid memory leak */
+
+		nsem->name = k_calloc(namelen + 1, sizeof(uint8_t));
+		if (nsem->name == NULL) {
+			errno = ENOSPC;
+			goto cleanup;
+		}
+
+		strcpy(nsem->name, name);
+		(void)atomic_set(&nsem->ref_count, 1);
+
+		if (sem_init(&nsem->sem, 0, value) != 0) {
+			errno = EINVAL;
+			goto cleanup;
+		}
+
+		k_mutex_lock(&nsem_mutex, K_FOREVER);
+		sys_slist_append(&sem_list, (sys_snode_t *)&(nsem->snode));
+		k_mutex_unlock(&nsem_mutex);
+	} else {
+		atomic_inc(&nsem->ref_count);
+	}
+
+	return &nsem->sem;
+
+cleanup:
+	cleanup_nsem(nsem);
+
+error:
+	return (sem_t *)SEM_FAILED;
+}
```
@npitre - would this address the raciness issue?

```suggestion
sem_t *sem_open(const char *name, int oflags, ...)
{
	va_list va;
	mode_t mode;
	unsigned int value;
	struct nsem_obj *nsem = NULL;
	size_t namelen;

	if (name == NULL) {
		errno = EINVAL;
		return NULL;
	}

	namelen = strlen(name);
	if ((namelen + 1) > CONFIG_SEM_NAMELEN_MAX) {
		errno = ENAMETOOLONG;
		return NULL;
	}

	/* Check if the named semaphore exists */
	k_mutex_lock(&nsem_mutex, K_FOREVER);
	nsem = find_nsem(name);

	if ((nsem != NULL) && (oflags & O_CREAT) != 0 &&
	    (oflags & O_EXCL) != 0) {
		/* Named semaphore has already been opened and O_EXCL is set */
		errno = EEXIST;
		goto error;
	}

	if (nsem != NULL) {
		atomic_inc(&nsem->ref_count);
		goto unlock;
	}

	if ((oflags & O_CREAT) == 0) {
		nsem = NULL;
		errno = ENOENT;
		goto error;
	}

	va_start(va, oflags);
	BUILD_ASSERT(sizeof(mode_t) <= sizeof(int));
	mode = va_arg(va, int);
	value = va_arg(va, unsigned int);
	va_end(va);

	nsem = k_malloc(sizeof(struct nsem_obj));
	if (nsem == NULL) {
		errno = ENOSPC;
		goto error;
	}

	/* goto `cleanup` past this point to avoid memory leak */

	nsem->name = strdup(name);
	if (nsem->name == NULL) {
		errno = ENOSPC;
		goto cleanup;
	}

	(void)atomic_set(&nsem->ref_count, 1);

	if (sem_init(&nsem->sem, 0, value) != 0) {
		errno = EINVAL;
		goto cleanup;
	}

	sys_slist_append(&sem_list, (sys_snode_t *)&(nsem->snode));
	goto unlock;

cleanup:
	cleanup_nsem(nsem);
	nsem = NULL;

unlock:
	k_mutex_unlock(&nsem_mutex);

	if (nsem == NULL) {
		return (sem_t *)SEM_FAILED;
	}
	
	return &nsem->sem;
}
```

ycsin replied:
I've refactored the code to lock the entire section before checking if a named semaphore exists:

> The check for the existence of the semaphore and the creation of the semaphore if it does not exist are atomic with respect to other processes executing sem_open() with O_EXCL and O_CREAT set

cfriedt commented on lib/posix/semaphore.c:
```diff
@@ -148,3 +202,160 @@ int sem_wait(sem_t *semaphore)
+int sem_unlink(const char *name)
+{
+	struct nsem_obj *nsem;
+	size_t namelen;
+
+	if (name == NULL) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	namelen = strlen(name);
+	if ((namelen + 1)  > CONFIG_SEM_NAMELEN_MAX) {
```
```suggestion
	if (name == NULL) {
		errno = EINVAL;
		return -1;
	}

	if ((strlen(name) + 1)  > CONFIG_SEM_NAMELEN_MAX) {
```

ycsin replied:
fixed

cfriedt commented on lib/posix/semaphore.c:
```diff
@@ -148,3 +202,160 @@ int sem_wait(sem_t *semaphore)
+		return -1;
+	}
+
+	k_free(nsem->name);
```
malloc / free would be fine to use here instead of `k_malloc()` / `k_free()`

ycsin replied:
Are they better? I used `k_` variant as that is what the `key.c` & `mqueue.c` implementations used

cfriedt replied:
`k_` variants are fine too (probably better than malloc, to be honest)

npitre commented:
The code now looks correct, the race is gone.

Here's some suggestions that aren't critical but could optimize the code
further:

- All `ref_count` manipulations, except for one case, happen while 
  `nsem_mutex` is taken. The exception is `sem_close()` which is unlikely 
  to be on any performance-critical path. By enlarging the `nsem_mutex`
  coverage in `sem_close()` you could drop all atomic operations and convert 
  `ref_count` to a plain int which would reduce the binary footprint, and
  possibly run faster on some architectures.

- You could consider the presence of a name as `ref_count` worthy. This 
  means initializing `ref_count` to 2 instead of 1 in `sem_open()` and
  decrementing `ref_count` in `sem_unlink()`. This would allow for removing 
  the `nsem->name == NULL` test in `sem_close()` relying on the count alone

- With the above, the `ref_count` decrements could be factored into
  `remove_nsem_if_unused()` (and the function properly renamed to something
  like e.g. nsem_put_ref() or the like) which would reduce binary footprint 
  further. This would open the possibility for cleanly asserting that the 
  count does not go negative and there is no longer a name if the count is 0 
  in only one place.


-

npitre commented:
And another comment. You have:

```
#define SEM_FAILED (-1)
```

Since this is used where a pointer is normally returned, it would be nicer 
to add the cast to the definition directly. Also, it is typically defined 
as 0 so the error case can be seen as a null pointer return. So I'd suggest:

```
#define SEM_FAILED ((sem_t *) 0)
```

ycsin commented:
> * You could consider the presence of a name as `ref_count` worthy. This
>   means initializing `ref_count` to 2 instead of 1 in `sem_open()` and
>   decrementing `ref_count` in `sem_unlink()`. This would allow for removing
>   the `nsem->name == NULL` test in `sem_close()` relying on the count alone
> * With the above, the `ref_count` decrements could be factored into
>   `remove_nsem_if_unused()` (and the function properly renamed to something
>   like e.g. nsem_put_ref() or the like) which would reduce binary footprint
>   further. This would open the possibility for cleanly asserting that the
>   count does not go negative and there is no longer a name if the count is 0
>   in only one place.

Hey @npitre, thanks for the suggestions!

Regarding to point 2 and 3, I've added assert tests to make sure that the `ref_count` doesn't underflow or overflow, but I'm not sure how to properly 'encodes' the unlink function call into the `ref_count` while still be able to guarantee that a named semaphore is destroyed only if it is unlinked and closed, i.e. without the `nsem->name == NULL` test, wouldn't `sem_close` + `sem_close` equals to `sem_unlink` + `sem_close`?

ycsin commented:
Updated to:
- adopt suggestions from @npitre (refactor for the removal of atomic, under/overflow tests)
- added a little stress test for the named semaphore (one thread `sem_open` a semaphore repeatedly and finally `sem_unlink`, the other thread doing `sem_close`)
- added a bit more testpoints for using a named semaphore in the normal semaphore test.

ycsin committed:
tests: posix: semaphore: refactor `test_semaphore`

Localize a few public variables and refactor the test and
functions so that they are reusable.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: posix: semaphore: run normal semaphore test with named semaphore

Run the normal semaphore test with a named semaphore.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

npitre committed:
posix: semaphore: optimize named semaphore implementation a bit

- Regroup refcount decrement and semaphore destruction by making the
  linked state into a counted reference for it. This allows for
  simplifying the code and cleanly adding a few assertions in a common
  location.

- Remove redundant initialization to NULL on memory about to be freed
  and local pointer in nsem_cleanup().

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
tests: posix: semaphore: assorted adjustments

- Adjust refcount checks with regards to previous commit.

- Remove redundant zassert_not_null() on local pointers after a
  sem_close(). There is no implicit reference passing in C.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre commented:
> i.e. without the `nsem->name == NULL` test, wouldn't `sem_close` +
> `sem_close` equals to `sem_unlink` + `sem_close`?

Consider these events:

```
operation                refcount change          refcount state
----------------------------------------------------------------

sem_open()               = 2 (created)            2
sem_open()               + 1 (already exists)     3
sem_close()              - 1                      2
sem_close()              - 1                      1
sem-unlink()             - 1                      0 (destroyed)

sem_open()               = 2 (created)            2
sem_open()               + 1 (already exists)     3
sem_unlink()             - 1                      2
sem_close()              - 1                      1
sem_close()              - 1                      0 (destroyed)
```

I'll push this change with a few misc tidbits on top of your pull request.
Feel free to fold them in your original commits or have them merged as is.

ycsin commented on lib/posix/semaphore.c:
```diff
@@ -53,17 +53,20 @@ static void nsem_cleanup(struct nsem_obj *nsem)
+	__ASSERT(nsem->ref_count >= 0, "ref_count may not be negative");
+
 	if (nsem->ref_count == 0) {
+		__ASSERT(nsem->name == NULL, "ref_count is 0 but sem is not unlinked");
```
@npitre This means that the following behavior is guaranteed only when `CONFIG_ASSERT` is enabled, is that intended?

> If the semaphore has not been removed with a successful call to [sem_unlink()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_unlink.html), then [sem_close()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_close.html) has no effect on the state of the semaphore.

i.e.: a named semaphore can be closed without being unlinked:

```
operation                refcount change          refcount state
----------------------------------------------------------------

sem_open()               = 2 (created)            2
sem_close()              - 1                      1
sem_close()              - 1                      0
```

npitre commented:
>>  if (nsem->ref_count == 0) {
>>+   __ASSERT(nsem->name == NULL, "ref_count is 0 but sem is not unlinked");
>
> @npitre This means that the following behavior is guaranteed only when
> `CONFIG_ASSERT` is enabled, is that intended?

Yes. This would happen only if the semaphore user's code is buggy.
This should never happen with well behaved code. Assertions can be configured
out to save on binary size and runtime overhead once the code is proven to 
behave properly.

>> If the semaphore has not been removed with a successful call to
>> sem_unlink()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_unlink.html),
>> then em_close()](https://pubs.opengroup.org/onlinepubs/9699919799/functions/sem_close.html)
>> has no effect on the state of the semaphore.
>
> i.e.: a named semaphore can be closed without being unlinked

Yes, that is what the code does. The test suite exercizes that condition too.

> ```
> operation                refcount change          refcount state
> ----------------------------------------------------------------
>
> sem_open()               = 2 (created)            2
> sem_close()              - 1                      1
> sem_close()              - 1                      0
> ```

That is a bad example. You normally must have balanced `open()` and
`close()`. Closing your own instance twice is buggy. Inagine what would
happen if you have an `open()` from two different threads and one thread
does a `close()` twice. The other thread is unlikely to be happy about that.

Multiple `close()` for a single `open()` is so unusual and unexpected
that the spec would mention it explicitly if that were allowed.

ycsin commented on lib/posix/semaphore.c:
```diff
@@ -15,7 +15,7 @@
 	sys_snode_t snode;
 	sem_t sem;
-	unsigned int ref_count;
+	int ref_count;
```
@npitre I've modified other part that uses `ref_count` to `int` as well, and changed the overflow test from `__ASSERT_NO_MSG(nsem->ref_count != UINT_MAX)` to `__ASSERT_NO_MSG(nsem->ref_count != INT_MAX)`

ycsin commented:
> Feel free to fold them in your original commits or have them merged as is.

@npitre I've took the liberty to fold some of the fixes from your commits into the original commits where they where initially introduced and reworded your commits, but left the `ref_count` changes as-is in your commits
